crates/cubecl-opt/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+use std::{
+    collections::{HashMap, VecDeque},
+    ops::{Deref, DerefMut},
+    rc::Rc,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+use cubecl_core::{
+    ir::{self as core, Operator, Procedure, Variable},
+    CubeDim,
+};
+use cubecl_core::{
+    ir::{Item, Operation, Scope},
+    ExecutionMode,
+};
+use passes::{
+    CompositeMerge, ConstEval, ConstOperandSimplify, CopyPropagateArray, CopyTransform,
+    EliminateConstBranches, EliminateDeadBlocks, EliminateUnusedVariables, FindConstSliceLen,
+    InBoundsToUnchecked, InlineAssignments, IntegerRangeAnalysis, MergeSameExpressions,
+    OptimizationPass, RemoveIndexScalar,
+};
+use petgraph::{prelude::StableDiGraph, visit::EdgeRef, Direction};
+
+mod block;
+mod control_flow;
+mod debug;
+mod instructions;
+mod passes;
+mod phi_frontiers;
+mod version;
+
+pub use block::*;
+pub use control_flow::*;
+pub use petgraph::graph::{EdgeIndex, NodeIndex};
+
+#[derive(Clone, Debug, Default)]
+pub struct AtomicCounter {
+    inner: Rc<AtomicUsize>,
+}
+
+impl AtomicCounter {
+    pub fn new(val: usize) -> Self {
+        Self {
+            inner: Rc::new(AtomicUsize::new(val)),
+        }
+    }
+
+    pub fn inc(&self) -> usize {
+        self.inner.fetch_add(1, Ordering::AcqRel)
+    }
+
+    pub fn get(&self) -> usize {
+        self.inner.load(Ordering::Acquire)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct Slice {
+    pub(crate) start: Variable,
+    pub(crate) end: Variable,
+    pub(crate) end_op: Option<Operation>,
+    pub(crate) const_len: Option<u32>,
+}
+
+#[derive(Default, Debug, Clone)]
+struct Program {
+    pub variables: HashMap<(u16, u8), Item>,
+    pub slices: HashMap<(u16, u8), Slice>,
+    pub graph: StableDiGraph<BasicBlock, ()>,
+    root: NodeIndex,
+    int_ranges: HashMap<VarId, Range>,
+}
+
+impl Deref for Program {
+    type Target = StableDiGraph<BasicBlock, ()>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.graph
+    }
+}
+
+impl DerefMut for Program {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.graph
+    }
+}
+
+type VarId = (u16, u8, u16);
+
+#[derive(Default, Clone, Copy, PartialEq, Eq, Debug)]
+struct Range {
+    lower_bound: Option<i64>,
+    upper_bound: Option<i64>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Optimizer {
+    program: Program,
+    pub current_block: Option<NodeIndex>,
```

Comment:
Is there a reason to have this as public? Is it mutated by each optimization then reset after? It's a bit weird to have a counter/ptr as a public member of the optimizer.